common/common.h

```diff
@@ -110,9 +111,16 @@ enum common_conversation_mode {
     COMMON_CONVERSATION_MODE_AUTO     = 2,
 };
 
+enum common_grammar_trigger_type {
+    COMMON_GRAMMAR_TRIGGER_TYPE_TOKEN,
+    COMMON_GRAMMAR_TRIGGER_TYPE_WORD,
+    COMMON_GRAMMAR_TRIGGER_TYPE_PATTERN,
+    COMMON_GRAMMAR_TRIGGER_TYPE_PATTERN_START,
+};
+
 struct common_grammar_trigger {
-    std::string word;
-    bool at_start;
+    common_grammar_trigger_type type;
+    std::variant<llama_token, std::string> value;
```

Comment:
A bit hesitating about the usage of `std::variant` here. Overall, it's OK, but it's the first time we use it in the codebase, and I'm not sure it is worth introducing it as a pattern. Having 2 separate members `value_token` and `value_string` is the alternative.

Reply:
I agree that the `std::variant` is a bit redundant here, for we can have 2 variables as @ggerganov suggested above, then set `value_token` to `LLAMA_TOKEN_NULL` if it is unused

Reply:
variant and option are out (for now ;-))